ShibMd:Scope matching should be case insensitive.

thijskh opened the pull request:
Target attributes are defined as being case insensitive, so any
matching of the scope on them should be the same.

thijskh committed:
ShibMd:Scope matching should be case insensitive.

Target attributes are defined as being case insensitive, so any
matching of the scope on them should be the same.

arothuis commented on src/OpenConext/Value/Saml/Metadata/ShibbolethMetadataScope.php:
```diff
@@ -66,7 +66,7 @@ public function allows($string)
 
         if (!$this->isRegexp) {
-            return $this->scope === $string;
+            return strcasecmp($this->scope, $string) === 0;
```
Nice use of `strcasecmp`.

arothuis reviewed (commented):
Looks good to me.

thijskh committed:
ShibMdScope pattern delimiters are not part of metadata.

According to https://wiki.shibboleth.net/confluence/display/SC/ShibMetaExt+V1.0,
in case of regexp being true, the value of ShibMd:Scope will be solely a
regexp pattern, not including any delimiters. The code should therefore
wrap this before passing it to RegularExpression.

This is a good thing because passing it as a complete regular expression
also allows one to set all modifiers, which is likely undesirable and
a security risk (e.g. by passing the "e" modifier).

thijskh committed:
Add i modifier to match regular expressions case insensitively.

This is to make the behaviour similar to the literal behaviour.

thijskh commented:
@arothuis

arothuis commented on src/OpenConext/Value/Saml/Metadata/ShibbolethMetadataScope.php:
```diff
@@ -66,10 +66,10 @@ public function allows($string)
         }
 
-        $regexp = new RegularExpression($this->scope);
+        $regexp = new RegularExpression("#$this->scope#i");
```
Nit: we tend to use sprintf and/or non-interpolated single quoted strings and concatenation for these kinds of things, e.g. `'#' . $this->scope . '#i'`.

arothuis commented:
As I understand it, this change makes the ShibMd value objects behave more according to the specification. I'm a proponent for these kinds of changes, but the changes regarding the removal of the regex delimiters might have more impact for dependents of this library that had not implemented the specification correctly. 

How does this align with how the scopes are being handled currently, for instance in OpenConext-engineblock / metadata?

thijskh commented:
The regexp part is not used in SURFconext (currently). If we had used it, we would have noticed that it was broken. Because metadata does not specify regexps with delimiters, it cannot have worked. The bugfix makes it start working. If anyone else had used it, they'd have noticed that it was not usable as-is.

As for the case insensitive matching: this makes more things match the scope, not less, so is highly unlikely to break something that was working before.

In any case, the specific functionality around scopes was introduced for EB, so I'm quite sure that no others are using it yet.

All in all, I think very safe to merge.

thijskh committed:
Coding style: avoid use of interpolated strings.

arothuis commented:
Thanks for the thorough explanation. Also, I appreciate the clear commit messages!

🚢

arothuis commented on src/OpenConext/Value/Saml/Metadata/ShibbolethMetadataScope.php:
```diff
@@ -50,7 +50,7 @@ public function __construct($scope, $isRegexp = false)
 
         if ($isRegexp) {
-            Assertion::validRegularExpression($scope, 'scope');
+            Assertion::validRegularExpression('#' . $scope . '#i', 'scope');
```
I'm not sure, but the [delimiters](http://php.net/manual/en/regexp.reference.delimiters.php) used here might introduce an opinion regarding how the regex should be escaped. For instance, using `#` will make it so that slashes, which are other common regex delimiters, do not need to be escaped. This might cause issues when people do escape their slashes.

I cannot find which type of regex (PCRE, POSIX, ERE, BRE) is being used in shibmd scopes and I also cannot find which escaping mechanism is preferred.

thijskh replied:
This is unfortunately rather undefined in the 'shibmb:scope' "spec". However, I think it's unlikely that the attribute would contain any actual slashes since it's supposed to contain DNS domain names. I think we can live with this. Regexps are discouraged in any case.

arothuis replied:
I have run some tests, escaping non-alphanumeric, non-backslash, non-whitespace characters will give the same result regardless of your delimiter (except if you use a special character as delimiter). So it does not cause any issues when people do escape their slashes.
